winch/codegen/src/isa/aarch64/asm.rs

```diff
@@ -397,6 +399,80 @@ impl Assembler {
         self.emit_alu_rrrr(ALUOp3::MAdd, scratch, rn, rd, regs::zero(), size);
     }
 
+    /// Signed/unsigned division with three registers.
+    pub fn div_rrr(
+        &mut self,
+        divisor: Reg,
+        dividend: Reg,
+        dest: Writable<Reg>,
+        kind: DivKind,
+        size: OperandSize,
+    ) {
+        // Check for division by 0
+        self.emit(Inst::TrapIf {
+            kind: CondBrKind::Zero(divisor.into()),
+            trap_code: TrapCode::INTEGER_DIVISION_BY_ZERO,
+        });
+
+        // check for overflow
+        if kind == DivKind::Signed {
+            // we first check whether the divisor is -1
+            self.emit(Inst::AluRRImm12 {
+                alu_op: ALUOp::AddS,
+                size: size.into(),
+                rd: writable!(zero().into()),
+                rn: divisor.into(),
+                imm12: Imm12::maybe_from_u64(1).expect("1 fits in 12 bits"),
+            });
+            // if it is -1, then we check if the dividend is MIN
+            self.emit(Inst::CCmpImm {
+                size: size.into(),
+                rn: dividend.into(),
+                imm: UImm5::maybe_from_u8(1).expect("1 fits in 5 bits"),
+                nzcv: NZCV::new(false, false, false, false),
+                cond: Cond::Eq,
+            });
+
+            // Finally, trap if the previous operation overflowed
+            self.emit(Inst::TrapIf {
+                kind: CondBrKind::Cond(Cond::Vs),
+                trap_code: TrapCode::INTEGER_OVERFLOW,
+            })
+        }
+
+        // cranelif-codegen doesn't support emitting u/sdiv for anything but I64,
```

Comment:
```suggestion
        // `cranelift-codegen` doesn't support emitting u/sdiv for anything but I64,
```

Reply:
Would you mind linking to this issue from this comment? https://github.com/bytecodealliance/wasmtime/issues/9766